# Analysis of Variance (ANOVA)

In [1]:
from data.create_data import *
import numpy as np
import pandas as pd
import scipy
from scipy import stats
import matplotlib.pyplot as plt

%matplotlib inline

In [2]:
data = read_frmgham()
data.educ.dropna(inplace=True)
data['educ'] = data.educ.astype('int')
data

randid     sex  totchol  age  sysbp  diabp  cursmoke  cigpday    bmi  \
0         2448    male    195.0   39  106.0   70.0         0      0.0  26.97   
1         2448    male    209.0   52  121.0   66.0         0      0.0    NaN   
2         6238  female    250.0   46  121.0   81.0         0      0.0  28.73   
3         6238  female    260.0   52  105.0   69.5         0      0.0  29.43   
4         6238  female    237.0   58  108.0   66.0         0      0.0  28.50   
5         9428    male    245.0   48  127.5   80.0         1     20.0  25.34   
6         9428    male    283.0   54  141.0   89.0         1     30.0  25.34   
7        10552  female    225.0   61  150.0   95.0         1     30.0  28.58   
8        10552  female    232.0   67  183.0  109.0         1     20.0  30.18   
9        11252  female    285.0   46  130.0   84.0         1     23.0  23.10   
10       11252  female    343.0   51  109.0   77.0         1     30.0  23.48   
11       11252  female      NaN   58  155.0   90.0         1     30.0  24.61   
12       11263  female    228.0   43  180.0  110.0         0      0.0  30.30   
13       11263  female    230.0   49  177.0  102.0         0      0.0  31.36   
14       11263  female    220.0   55  180.0  106.0         0      0.0  31.17   
15       12629  female    205.0   63  138.0   71.0         0      0.0  33.11   
16       12629  female    220.0   70  149.0   81.0         0      0.0  36.76   
17       12806  female    313.0   45  100.0   71.0         1     20.0  21.68   
18       12806  female      NaN   51  109.5   72.5         1     30.0  22.19   
19       12806  female    320.0   57  110.0   46.0         1     30.0  22.02   
20       14367    male    260.0   52  141.5   89.0         0      0.0  26.36   
21       14367    male    292.0   58  132.0   90.0         0      0.0  25.39   
22       14367    male    280.0   64  168.0  100.0         0      0.0  25.72   
23       16365    male    225.0   43  162.0  107.0         1     30.0  23.61   
24       16365    male    258.0   49  147.0  102.0         0      0.0  27.50   
25       16365    male    211.0   55  173.0  123.0         0      0.0  29.11   
26       16799  female    254.0   50  133.0   76.0         0      0.0  22.91   
27       16799  female      NaN   56  127.5   70.0         0      0.0  20.30   
28       16799  female    291.0   62  120.0   70.0         0      0.0  21.98   
29       19304  female    247.0   43  131.0   88.0         0      0.0  27.64   
...        ...     ...      ...  ...    ...    ...       ...      ...    ...   
11597  9969773  female    229.0   62  188.0   91.0         0      0.0  34.74   
11598  9973350  female    251.0   51  140.0   80.0         1     20.0  25.60   
11599  9973350  female    267.0   57  145.0   79.0         1     20.0  22.14   
11600  9973350  female    267.0   63  160.0   88.0         0      0.0  25.60   
11601  9978986  female    268.0   56  170.0  102.0         1      3.0  22.89   
11602  9978986  female    280.0   62  133.0   82.0         1      5.0  22.54   
11603  9978986  female    243.0   68  110.0   66.0         0      0.0  24.30   
11604  9982118    male    187.0   58  141.0   81.0         0      0.0  24.96   
11605  9982118    male    256.0   64  156.5   82.5         0      0.0  25.56   
11606  9982118    male    219.0   70  163.5   75.0         0      0.0  25.26   
11607  9983319    male    176.0   68  168.0   97.0         0      0.0  23.14   
11608  9984683    male    313.0   50  179.0   92.0         1      1.0  25.97   
11609  9984683    male      NaN   56  157.5   89.0         1      1.0  24.85   
11610  9989287    male    207.0   51  126.5   80.0         1     43.0  19.71   
11611  9989287    male    210.0   57  116.0   69.0         1     30.0  17.65   
11612  9989287    male      NaN   63  120.0   80.0         1     50.0  17.38   
11613  9990894  female    248.0   48  131.0   72.0         1     20.0  22.00   
11614  9990894  female    258.0   54  145.0   89.0         1     20.0  21.64   
11615  9

### Research Question
1. Is there any difference in the mean blood pressures across the various levels of education? 
  * Do any of the group means differ from one another?
  * Can a significant proportion of the overall variability found in the blood pressures be attributed to the known differences between the groups?  

#### Legend: Factor/Treatment = `Education Level`
The group or categorical variable is `educ`, which persists of 4 levels:
  * **`1`**: 0-11 years of education (denoted as `elem`)
  * **`2`**: High School Diploma, GED (denoted as `high`)
  * **`3`**: some college, vocational school (denoted as `some_college`)
  * **`4`**: College degree or more (denoted as `college`)

## One-way ANOVA
One-way ANOVA tests whether the mean of systolic or diastolic blood pressure differs across the various levels of education.

The ANOVA test enables comparison of more than two group means, while keeping the Type I error at 0.05.

### Test-statistic: `F-statistic`
The **F-statistic** or **F-ratio** is the ratio comparing variance.  

`F = variance`<sub>between groups</sub>` / variance`<sub>within groups</sub>

Interpretations:
  * **F = 1.0**: There is no difference in the mean blood pressures of individuals in the various education groups.
    * H<sub>0</sub> is true
  * **F > 1.0**: There is a difference in the mean blood pressures of individuals in the various education groups.
    * H<sub>0</sub> is false

#### Systolic Blood Pressure

1) Hypotheses
  * **H<sub>0</sub>**: The sample mean of systolic blood pressure points to a population where the population mean is μ so that μ<sub>1</sub> = μ<sub>2</sub> = μ<sub>3</sub> = μ<sub>4</sub>.
  * **H<sub>A</sub>**: The sample mean of systolic blood pressure points to a population where the population mean is not μ so that μ<sub>1</sub> ≠  μ<sub>2</sub> ≠ μ<sub>3</sub> ≠ μ<sub>4</sub>.
  
  
2) Compute **t-statistic** (F-statistic/F-ratio) & **p-value**.

In [3]:
elem_sysbp = data[data['educ'] == 1]['sysbp']
high_sysbp = data[data['educ'] == 2]['sysbp']
some_college_sysbp = data[data['educ'] == 3]['sysbp']
college_sysbp = data[data['educ'] == 4]['sysbp']

In [4]:
f_sysbp, fpval_sysbp = stats.f_oneway(elem_sysbp, high_sysbp, 
                                      some_college_sysbp, college_sysbp)
print "One-way ANOVA F-statistic = %.2f" % f_sysbp
print "One-way ANOVA p-value = %.2f" % fpval_sysbp

One-way ANOVA F-statistic = 55.53
One-way ANOVA p-value = 0.00


3) Results
  * test-statistic: **F-statistic** = 55.53
  * **p-value** = 0

The **p-value** is small (p-value < 0.05), providing substantial evidence against the null hypothesis (H<sub>0</sub>). Hence, H<sub>0</sub> is rejected in favor of the alternative hypothesis (H<sub>A</sub>).

Additionally, the test-statistic, **F-statistic/ratio** is greater than 1. Hence, further supporting the rejection of H<sub>0</sub>.

##### Conclusion
μ<sub>1</sub> ≠  μ<sub>2</sub> ≠ μ<sub>3</sub> ≠ μ<sub>4</sub>

Based on the results, it's concluded that there is a statistically significant difference in the mean systolic blood pressure (mmHg) of individuals in the various levels of education (groups 1, 2, 3, 4). The mean systolic blood pressure vary between the groups.

#### Diastolic Blood Pressure

1) Hypotheses
  * **H<sub>0</sub>**: The sample mean of diastolic blood pressure points to a population where the population mean is μ so that μ<sub>1</sub> = μ<sub>2</sub> = μ<sub>3</sub> = μ<sub>4</sub>.
  * **H<sub>A</sub>**: The sample mean of diastolic lbood pressure points to a population where the population mean is not μ so that μ<sub>1</sub> ≠  μ<sub>2</sub> ≠ μ<sub>3</sub> ≠ μ<sub>4</sub>.
  
  
2) Compute **t-statistic** (F-statistic/F-ratio) & **p-value**.

In [5]:
elem_dbp = data[data['educ'] == 1]['diabp']
high_dbp = data[data['educ'] == 2]['diabp']
some_college_dbp = data[data['educ'] == 3]['diabp']
college_dbp = data[data['educ'] == 4]['diabp']

In [6]:
f_diabp, fpval_diabp = stats.f_oneway(elem_dbp, high_dbp, 
                                      some_college_dbp, college_dbp)
print "One-way ANOVA F-statistic = %.2f" % f_diabp
print "One-way ANOVA p-value = %.2f" % fpval_diabp

One-way ANOVA F-statistic = 8.91
One-way ANOVA p-value = 0.00


3) Results
  * test-statistic: **F-statistic** = 8.91
  * **p-value** = 0

The **p-value** is small (p-value < 0.05), providing substantial evidence against the null hypothesis (H<sub>0</sub>). Hence, H<sub>0</sub> is rejected in favor of the alternative hypothesis (H<sub>A</sub>).

Additionally, the test-statistic, **F-statistic/ratio** is greater than 1. Hence, further supporting the rejection of H<sub>0</sub>.

##### Conclusion
μ<sub>1</sub> ≠  μ<sub>2</sub> ≠ μ<sub>3</sub> ≠ μ<sub>4</sub>

Based on the results, it's concluded that there is a statistically significant difference in the mean diastolic blood pressure (mmHg) of individuals in the various levels of education (groups 1, 2, 3, 4). The mean diatolic blood pressure vary between the groups.